static/src/javascripts/projects/common/modules/commercial/article-body-adverts.js

```diff
@@ -31,97 +43,114 @@ define([
-        init = function () {
-            if (!commercialFeatures.articleBodyAdverts) {
-                return false;
-            }
+    // If a merchandizing component has been rendered but is empty,
+    // we allow a second pass for regular inline ads. This is because of
+    // the decoupling between the spacefinder algorightm and the targeting
+    // in DFP: we can only know if a slot can be removed after we have
+    // received a response from DFP
+    function onAdRendered(event) {
+        if (event.slot.getSlotElementId() === 'dfp-ad--im' && event.isEmpty) {
+            mediator.off('modules:commercial:dfp:rendered', onAdRendered);
+            addArticleAds(2, getRules()).then(function (countAdded) {
+                return countAdded === 2 ?
+                    addArticleAds(8, getLongArticleRules()) :
+                    countAdded;
+            }).then(function () {
+                $('.ad-slot--inline').each(dfp.addSlot);
+            });
+        }
+    }
 
-            var rules = getRules();
+    function init() {
+        if (!commercialFeatures.articleBodyAdverts) {
+            return false;
+        }
 
-            if (config.page.hasInlineMerchandise) {
-                spaceFiller.fillSpace(getInlineMerchRules(), function (paras) {
-                    adNames.unshift(['im', 'im']);
-                    insertAdAtPara(paras);
-                });
-            }
+        var rules = getRules();
 
-            if (config.switches.viewability && detect.getBreakpoint() !== 'mobile') {
-                return tryAddingAdvert().then(tryAddingAdvert).then(function () {
-                    return addLongArticleAds(8);
-                });
-            } else {
-                return tryAddingAdvert().then(function () {
-                    if (detect.isBreakpoint({max: 'tablet'})) {
-                        return tryAddingAdvert();
-                    } else {
-                        return Promise.resolve(null);
-                    }
-                });
-            }
+        boot();
 
-            function tryAddingAdvert() {
-                return spaceFiller.fillSpace(rules, insertAdAtPara);
-            }
-        };
+        if (config.page.hasInlineMerchandise) {
+            addInlineMerchAd(getInlineMerchRules());
+        }
 
-    return {
-        init: init,
-        // rules exposed for spacefinder debugging
-        getRules: getRules,
-        getInlineMerchRules: getInlineMerchRules,
+        if (config.switches.viewability && detect.getBreakpoint() !== 'mobile') {
+            return addArticleAds(2, rules).then(function (countAdded) {
+                if (countAdded === 0) {
+                    mediator.on('modules:commercial:dfp:rendered', onAdRendered);
+                }
 
-        reset: function () {
-            ads = [];
+                return countAdded === 2 ?
+                    addArticleAds(8, getLongArticleRules()) :
```

Comment:
It might be useful for our future selves to document more what these magic count parameters mean.